File: src/main/java/catchla/yep/view/RecyclerViewBackport.java
/*
 * Copyright (c) 2015. Catch Inc,
 */

package catchla.yep.view;

import android.content.Context;
import android.support.v4.view.MotionEventCompat;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.InputDevice;
import android.view.MotionEvent;

import catchla.yep.util.MouseScrollDirectionDecider;


/**
 * Created by mariotaku on 15/3/30.
 */
public class RecyclerViewBackport extends RecyclerView {

    private final MouseScrollDirectionDecider mMouseScrollDirectionDecider;
    // This value is used when handling generic motion events.
    private float mScrollFactor = Float.MIN_VALUE;

    public RecyclerViewBackport(Context context) {
        this(context, null);
    }

    public RecyclerViewBackport(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RecyclerViewBackport(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        mMouseScrollDirectionDecider = new MouseScrollDirectionDecider(context, getScrollFactorBackport());
    }

    @Override
    public boolean onGenericMotionEvent(MotionEvent event) {
        final LayoutManager lm = getLayoutManager();
        if (lm == null) {
            return false;
        }
        if ((event.getSource() & InputDevice.SOURCE_CLASS_POINTER) != 0) {
            if (event.getAction() == MotionEventCompat.ACTION_SCROLL) {
                final float vScroll, hScroll;
                if (lm.canScrollVertically()) {
                    vScroll = event.getAxisValue(MotionEvent.AXIS_VSCROLL);
                    if (!mMouseScrollDirectionDecider.isVerticalAvailable()) {
                        mMouseScrollDirectionDecider.guessDirection(event);
                    }
                } else {
                    vScroll = 0f;
                }
                if (lm.canScrollHorizontally()) {
                    hScroll = event.getAxisValue(MotionEvent.AXIS_HSCROLL);
                    if (!mMouseScrollDirectionDecider.isHorizontalAvailable()) {
                        mMouseScrollDirectionDecider.guessDirection(event);
                    }
                } else {
                    hScroll = 0f;
                }
                if ((vScroll != 0 || hScroll != 0)) {
                    final float scrollFactor = getScrollFactorBackport();
                    float horizontalDirection = mMouseScrollDirectionDecider.getHorizontalDirection();
                    float verticalDirection = mMouseScrollDirectionDecider.getVerticalDirection();
                    final float hFactor = scrollFactor * (horizontalDirection != 0 ? horizontalDirection : -1);
                    final float vFactor = scrollFactor * (verticalDirection != 0 ? verticalDirection : -1);
                    smoothScrollBy((int) (hScroll * hFactor), (int) (vScroll * vFactor));
                }
            }
        }
        return false;
    }

    /**
     * Ported from View.getVerticalScrollFactor.
     */
    private float getScrollFactorBackport() {
        if (mScrollFactor == Float.MIN_VALUE) {
            TypedValue outValue = new TypedValue();
            if (getContext().getTheme().resolveAttribute(
                    android.R.attr.listPreferredItemHeight, outValue, true)) {
                mScrollFactor = outValue.getDimension(
                        getContext().getResources().getDisplayMetrics());
            } else {
                return 0; //listPreferredItemHeight is not defined, no generic scrolling
            }

        }
        return mScrollFactor;
    }

}

File: src/main/java/catchla/yep/view/RecyclerViewBackport.kt
/*
 * Copyright (c) 2015. Catch Inc,
 */

package catchla.yep.view

import android.content.Context
import android.support.v4.view.MotionEventCompat
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.util.TypedValue
import android.view.InputDevice
import android.view.MotionEvent

import catchla.yep.util.MouseScrollDirectionDecider


/**
 * Created by mariotaku on 15/3/30.
 */
class RecyclerViewBackport @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : RecyclerView(context, attrs, defStyle) {

    private val mouseScrollDirectionDecider: MouseScrollDirectionDecider
    // This value is used when handling generic motion events.
    private var scrollFactor = Float.NaN

    init {
        mouseScrollDirectionDecider = MouseScrollDirectionDecider(context, scrollFactorBackport)
    }

    override fun onGenericMotionEvent(event: MotionEvent): Boolean {
        val lm = layoutManager ?: return false
        if (event.source and InputDevice.SOURCE_CLASS_POINTER != 0) {
            if (event.action == MotionEventCompat.ACTION_SCROLL) {
                val vScroll: Float
                val hScroll: Float
                if (lm.canScrollVertically()) {
                    vScroll = event.getAxisValue(MotionEvent.AXIS_VSCROLL)
                    if (!mouseScrollDirectionDecider.isVerticalAvailable) {
                        mouseScrollDirectionDecider.guessDirection(event)
                    }
                } else {
                    vScroll = 0f
                }
                if (lm.canScrollHorizontally()) {
                    hScroll = event.getAxisValue(MotionEvent.AXIS_HSCROLL)
                    if (!mouseScrollDirectionDecider.isHorizontalAvailable) {
                        mouseScrollDirectionDecider.guessDirection(event)
                    }
                } else {
                    hScroll = 0f
                }
                if (vScroll != 0f || hScroll != 0f) {
                    val scrollFactor = scrollFactorBackport
                    val horizontalDirection = mouseScrollDirectionDecider.horizontalDirection
                    val verticalDirection = mouseScrollDirectionDecider.verticalDirection
                    val hFactor = if (horizontalDirection != 0) scrollFactor * horizontalDirection else -1f
                    val vFactor = if (verticalDirection != 0) scrollFactor * verticalDirection else -1f
                    smoothScrollBy((hScroll * hFactor).toInt(), (vScroll * vFactor).toInt())
                }
            }
        }
        return false
    }

    /**
     * Ported from View.getVerticalScrollFactor.
     */
    private //listPreferredItemHeight is not defined, no generic scrolling
    val scrollFactorBackport: Float
        get() {
            if (scrollFactor.isNaN()) {
                val outValue = TypedValue()
                if (context.theme.resolveAttribute(android.R.attr.listPreferredItemHeight, outValue, true)) {
                    scrollFactor = outValue.getDimension(context.resources.displayMetrics)
                } else {
                    return 0f
                }
            }
            return scrollFactor
        }

}
